### PR TITLE
Implementing metrics caching

### PR DESCRIPTION
**Pull request by janboll**
Related issue: https://issues.redhat.com/browse/APPSRE-6250

### CODE REVIEW

**Comment by mrWinston** [comment]
Hey @janboll !
Are you planning to do something about https://github.com/app-sre/aws-resource-exporter/blob/master/proxy.go ? @cblecker  mentioned yesterday that the caching functionality had already been implemented as part of the rds exporter in the past. i think it'd made sense to combine these two into one solution so we don't end up with two things implementing the same functionality

File: route53.go [hunk 1]
```diff
@@ -85,37 +82,30 @@ func (e *Route53Exporter) getRecordsPerHostedZoneMetrics(client *route53.Route53
...
-		return nil, err
+		return err
 	}
 
-	result = append(result,
-		prometheus.MustNewConstMetric(e.HostedZonesPerAccountQuota, prometheus.GaugeValue, quota),
-		prometheus.MustNewConstMetric(e.HostedZonesPerAccountUsage, prometheus.GaugeValue, float64(len(hostedZones))),
-	)
-	return result, nil
+	e.cache.AddMetric(prometheus.MustNewConstMetric(e.HostedZonesPerAccountQuota, prometheus.GaugeValue, quota))
```
[the remark this answers is not in the document]
Comment (janboll) [reply route53.go:106]: Good point! In the current state I just went ahead in a search and replace way to reduce the number of changes in the code.

File: main.go [hunk 2]
```diff
@@ -48,13 +48,15 @@ type BaseConfig struct {
 
 type RDSConfig struct {
 	BaseConfig `yaml:"base,inline"`
-	Regions    []string `yaml:"regions"`
+	Regions    []string      `yaml:"regions"`
+	Interval   time.Duration `yaml:"interval"`
```
Comment (bergmannf) [inline main.go:52]: This is probably something all collectors have, so you can add the `Interval` into the `BaseConfig` only once and it is included in every `Config` that will include it (which all configs should).

File: pkg/cache.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+func (mc *MetricsCache) GetAllMetrics() []prometheus.Metric {
+	mc.cacheMutex.Lock()
+	returnArr := make([]prometheus.Metric, 0)
+	for k, v := range mc.entries {
+		if time.Since(v.creation).Seconds() > float64(mc.ttlInSeconds) {
+			delete(mc.entries, k)
+		} else {
+			returnArr = append(returnArr, v.metric)
+			mc.entries[k] = v
```
Comment (bergmannf) [inline pkg/cache.go:48]: I think this is superfluous - the entries are already in the map, and I don't think appending them to the return slice will remove them.

File: main.go (hunk 2 above)
Comment (janboll) [reply main.go:52]: Added it to every config, cause some might have a lower interval, cause they need less time for collection.

File: pkg/cache.go (hunk 3 above)
Comment (janboll) [reply pkg/cache.go:48]: Good spot, that's a relic from a previous attempt. Will remove it!

File: main.go (hunk 2 above)
Comment (bergmannf) [reply main.go:52]: That should still work, as the baseconfig is just merged in - so every config has the field and can set it to a different value.

Comment (janboll) [reply main.go:52]: Missed that, thanks for pointing it out! I"ll update the PR.

File: route53.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrWinston) [reply route53.go:108]: Yes, that's totally fine for me. we can refactor that when we actually test the metric collection.

**Comment by bergmannf** [comment]
I ran the current code locally and having different labels will lead to only the latest metric being stored in the cache, because the key-hash does not have the actual values for the labels:

```
func getMetricHash(metric prometheus.Metric) string {
	fmt.Printf("Generating hash for %v\n", metric.)
	checksum := sha256.Sum256([]byte(metric.Desc().String()))
	return fmt.Sprintf("%x", checksum[:])
}
```

This outputs (as an excerpt):

```
Generating hash for Desc{fqName: "aws_resources_exporter_vpc_interfacevpcendpointspervpc_usage", help: "The usage of interface vpc endpoints per vpc", constLabels: {}, variableLabels: [aws_region vpcid]}
Generating hash for Desc{fqName: "aws_resources_exporter_vpc_interfacevpcendpointspervpc_usage", help: "The usage of interface vpc endpoints per vpc", constLabels: {}, variableLabels: [aws_region vpcid]}
Generating hash for Desc{fqName: "aws_resources_exporter_vpc_interfacevpcendpointspervpc_usage", help: "The usage of interface vpc endpoints per vpc", constLabels: {}, variableLabels: [aws_region vpcid]}
Generating hash for Desc{fqName: "aws_resources_exporter_vpc_interfacevpcendpointspervpc_usage", help: "The usage of interface vpc endpoints per vpc", constLabels: {}, variableLabels: [aws_region vpcid]}
Generating hash for Desc{fqName: "aws_resources_exporter_vpc_interfacevpcendpointspervpc_usage", help: "The usage of interface vpc endpoints per vpc", constLabels: {}, variableLabels: [aws_region vpcid]}
```

Which means it will loose metrics with different labels right now.

File: pkg/cache.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+func NewMetricsCache(ttlInSeconds int) *MetricsCache {
+	return &MetricsCache{
+		cacheMutex:   &sync.Mutex{},
+		entries:      map[string]CacheEntry{},
+		ttlInSeconds: ttlInSeconds,
+	}
+}
+
+func getMetricHash(metric prometheus.Metric) string {
+	checksum := sha256.Sum256([]byte(metric.Desc().String()))
```
Comment (mrWinston) [inline pkg/cache.go:27]: I am afraid we need to change the hash function a bit.
This only takes into account the static labels added during creation of the `metric.Desc`. Dynamic labels like `aws_region` are not respected when the hash is generated leading to those overwriting themselves.

You can add the following test that replicates that behaviour:

```
func TestSameMetricWithDifferentLabelsDontOverwrite(t *testing.T) {
	cache := NewMetricsCache(1)
	desc := prometheus.NewDesc("test", "multimetric", []string{"aws_region"}, nil)

	metricEast1 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1, "us-east-1")
	metricWest1 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1, "us-west-1")
	metricEast2 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 2, "us-east-1")
	cache.AddMetric(metricEast1)
	cache.AddMetric(metricWest1) // should *not* overwrite metricEast1
	cache.AddMetric(metricEast2) // should overwrite metricEast1

	assert.Len(t, cache.GetAllMetrics(), 2)
	assert.NotContains(t, cache.GetAllMetrics(), metricEast1)
	assert.Contains(t, cache.GetAllMetrics(), metricWest1)
	assert.Contains(t, cache.GetAllMetrics(), metricEast2)
}
```

Comment (janboll) [reply pkg/cache.go]: Good catch! thank's for testing @mrWinston and @bergmannf 
Perhaps it's possible to use the Write method for exporting a hashable representation?

Comment (janboll) [reply pkg/cache.go]: Hm, for me GetLabel gives me these:
```
metric.Write(&dto)
fmt.Println(dto.GetLabel())
output;
[name:"aws_region" value:"us-east-1" ]
```

Comment (janboll) [reply pkg/cache.go]: I pushed a commit, that should fix the hashing, please take a look. 
The string generated for hashing now looks like this:

```
Desc{fqName: "test", help: "multimetric", constLabels: {}, variableLabels: [aws_region]},aws_region,us-east-1
```

File: vpc.go [hunk 5]
```diff
@@ -68,65 +71,76 @@ func NewVPCExporter(sess []*session.Session, logger log.Logger, timeout time.Dur
...
-	c.collectVpcsPerRegionUsage(ch)
-	c.collectRoutesTablesPerVpcQuota(ch)
-	c.collectInterfaceVpcEndpointsPerVpcQuota(ch)
-	c.collectSubnetsPerVpcQuota(ch)
-	c.collectIPv4BlocksPerVpcQuota(ch)
-	vpcCtx, vpcCancel := context.WithTimeout(context.Background(), c.e.timeout)
+	ec2Svc := ec2.New(session)
+	quotaSvc := servicequotas.New(session)
+
+	e.collectVpcsPerRegionQuota(quotaSvc, *region)
```
[the remark this answers is not in the document]
Comment (janboll) [reply vpc.go:86]: Metrics look good

File: pkg/cache.go (hunk 4 above)
Comment (janboll) [reply pkg/cache.go]: I'll resolve this thread, since the test you provided looks good.

**Comment by janboll** [comment]
Compared the metrics from running the exporter using the main branch and this branch on one of our APP-SRE AWS accounts and did see no diff in the results (except for metric values, which is expected)

**Comment by janboll** [comment]
Done, I changed timeout to be optinal and added the default values to the Readme

**Review by mrWinston — APPROVED** [review]

**Comment by janboll** [comment]
Bypassing branch protection since @mrWinston is listed as approver https://github.com/app-sre/aws-resource-exporter/blob/master/OWNERS#L15

**Pull request merged**